app/cloudinit/cloudinit.go

```diff
@@ -559,39 +637,40 @@ var amazonLinuxTemplate = template.Must(template.New(oshelp.OSLinux).Funcs(funcs
 // Linux creates a userdata file for the Linux operating system.
 func Linux(params *Params) (payload string) {
 	sb := &strings.Builder{}
-
 	caCertPath := filepath.Join(certsDir, "ca-cert.pem")
 	certPath := filepath.Join(certsDir, "server-cert.pem")
 	keyPath := filepath.Join(certsDir, "server-key.pem")
+	templateData := struct {
+		Params
+		CaCertPath string
+		CertPath   string
+		KeyPath    string
+	}{
+		Params:     *params,
+		CaCertPath: caCertPath,
+		CertPath:   certPath,
+		KeyPath:    keyPath,
+	}
 	switch params.Platform.OSName {
 	case oshelp.AmazonLinux:
-		err := amazonLinuxTemplate.Execute(sb, struct {
-			Params
-			CaCertPath string
-			CertPath   string
-			KeyPath    string
-		}{
-			Params:     *params,
-			CaCertPath: caCertPath,
-			CertPath:   certPath,
-			KeyPath:    keyPath,
-		})
+		err := amazonLinuxTemplate.Execute(sb, templateData)
 		if err != nil {
 			panic(err)
 		}
 	default:
 		// Ubuntu
-		err := ubuntuTemplate.Execute(sb, struct {
-			Params
-			CaCertPath string
-			CertPath   string
-			KeyPath    string
-		}{
-			Params:     *params,
-			CaCertPath: caCertPath,
-			CertPath:   certPath,
-			KeyPath:    keyPath,
-		})
+		var err error
+		if params.GitspaceAgentConfig.VMInitScript == "" {
+			err = ubuntuTemplate.Execute(sb, templateData)
+		} else {
+			decodedScript, decodeErr := base64.StdEncoding.DecodeString(params.GitspaceAgentConfig.VMInitScript)
+			if decodeErr != nil {
+				err = fmt.Errorf("failed to decode the gitspaces vm init script: %w", err)
+				panic(err)
```

Comment:
don't panic, return appropriate errors (I know we are doing it in other places but we shouldn't ideally)

Reply:
done